File: unity/Assets/Scripts/Core/SaveManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;
using UnityEngine;

namespace KunchengRPG.Core
{
    /// <summary>
    /// Persists the save document to disk under Application.persistentDataPath.
    ///
    /// The JS original used localStorage with a quota; a file has no practical size
    /// limit, but the compression pass is kept so save size stays bounded across
    /// hundreds of generations and load stays fast.
    /// </summary>
    public class SaveManager
    {
        private const string FileName = "kuncheng_rpg_save.json";
        private const string LegacyFileName = "shenzhen_rpg_save.json";
        private const int FamilyLogMax = 200;
        private const int NpcMemoryMax = 500;
        private const int NpcEventMax = 20;

        public Data.SaveData Current { get; private set; }

        private static string SaveDir => Application.persistentDataPath;
        private static string SavePath => Path.Combine(SaveDir, FileName);
        private static string LegacyPath => Path.Combine(SaveDir, LegacyFileName);

        private bool migrated;

        /// <summary>
        /// Rename a pre-rebrand save into place, once per session.
        /// </summary>
        private void EnsureMigrated()
        {
            if (migrated) return;
            migrated = true;

            try
            {
                if (File.Exists(LegacyPath) && !File.Exists(SavePath))
                {
                    File.Move(LegacyPath, SavePath);
                    Debug.Log("[SaveManager] Migrated legacy save file");
                }
            }
            catch (Exception e)
            {
                Debug.LogWarning($"[SaveManager] Legacy save migration skipped: {e.Message}");
            }
        }

        public bool HasSave()
        {
            EnsureMigrated();
            return File.Exists(SavePath);
        }

        public bool Save(Data.SaveData state)
        {
            if (state == null) return false;
            EnsureMigrated();

            try
            {
                Compress(state);
                Directory.CreateDirectory(SaveDir);

                // Write to a temp file then swap, so a crash mid-write cannot
                // leave a truncated save where a valid one used to be.
                string tmp = SavePath + ".tmp";
                File.WriteAllText(tmp, JsonConvert.SerializeObject(state, Formatting.Indented));
                if (File.Exists(SavePath)) File.Delete(SavePath);
                File.Move(tmp, SavePath);

                Current = state;
                return true;
            }
            catch (Exception e)
            {
                Debug.LogError($"[SaveManager] Save failed: {e.Message}");
                return false;
            }
        }

        public Data.SaveData Load()
        {
            EnsureMigrated();
            if (!File.Exists(SavePath)) return null;

            try
            {
                var data = JsonConvert.DeserializeObject<Data.SaveData>(File.ReadAllText(SavePath));
                Current = data;
                return data;
            }
            catch (Exception e)
            {
                Debug.LogError($"[SaveManager] Load failed: {e.Message}");
                return null;
            }
        }

        public void DeleteSave()
        {
            try
            {
                if (File.Exists(SavePath)) File.Delete(SavePath);
                Current = null;
            }
            catch (Exception e)
            {
                Debug.LogWarning($"[SaveManager] Delete failed: {e.Message}");
            }
        }

        /// <summary>
        /// Absolute path to the save file, for "reveal in explorer" style UI.
        /// </summary>
        public string GetSavePath() => SavePath;

        public Data.SaveData CreateNewSave(string districtId = null, string originId = null)
        {
            var save = new Data.SaveData
            {
                version = 2,
                createdAtUnix = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                seed = UnityEngine.Random.Range(0, 999999),
                generation = 1,
                currentDistrictId = districtId,
                currentOriginId = originId
            };
            Current = save;
            return save;
        }

        /// <summary>
        /// Trim unbounded history in place: keep the most recent family log entries
        /// (summarising the rest), cap per-NPC event lists, and drop the least
        /// familiar NPCs once the roster gets large.
        /// </summary>
        private void Compress(Data.SaveData state)
        {
            if (state.familyLog != null && state.familyLog.Count > FamilyLogMax)
            {
                int dropCount = state.familyLog.Count - FamilyLogMax;
                var dropped = state.familyLog.GetRange(0, dropCount);

                state.familyLogSummary = Summarize(dropped, state.familyLogSummary);
                state.familyLog.RemoveRange(0, dropCount);
            }

            if (state.npcMemories == null || state.npcMemories.Count == 0) return;

            foreach (var mem in state.npcMemories.Values)
            {
                if (mem?.events != null && mem.events.Count > NpcEventMax)
                    mem.events.RemoveRange(0, mem.events.Count - NpcEventMax);
            }

            if (state.npcMemories.Count > NpcMemoryMax)
            {
                var ids = new List<string>(state.npcMemories.Keys);
                ids.Sort((a, b) => state.npcMemories[b].familiarity.CompareTo(state.npcMemories[a].familiarity));

                var kept = new Dictionary<string, Game.NpcMemory>(NpcMemoryMax);
                for (int i = 0; i < NpcMemoryMax; i++)
                    kept[ids[i]] = state.npcMemories[ids[i]];

                state.npcMemories = kept;
            }
        }

        /// <summary>
        /// Folds newly dropped entries into any existing summary rather than
        /// replacing it, so the running total survives repeated compression.
        /// </summary>
        private Data.FamilyLogSummary Summarize(
            List<Data.FamilyLogEntry> dropped, Data.FamilyLogSummary existing)
        {
            var summary = existing ?? new Data.FamilyLogSummary();
            summary.count += dropped.Count;

            foreach (var e in dropped)
                if (!summary.generations.Contains(e.generation))
                    summary.generations.Add(e.generation);

            summary.preview.Clear();
            for (int i = Math.Max(0, dropped.Count - 3); i < dropped.Count; i++)
                summary.preview.Add(dropped[i].title);

            return summary;
        }
    }
}

File: src/engine/core/SaveManager.js
const SAVE_KEY = 'kuncheng_rpg_save';
const OLD_SAVE_KEY = 'shenzhen_rpg_save';
const FAMILY_LOG_MAX = 200;
const NPC_MEMORY_MAX = 500;

export class SaveManager {
    constructor() {
        this.currentSave = null;
        this._migrated = false;
        this._storageAvailable = null; // null = not tested yet
    }

    _checkStorage() {
        if (this._storageAvailable !== null) return this._storageAvailable;
        try {
            const testKey = '__kuncheng_test__';
            localStorage.setItem(testKey, '1');
            localStorage.removeItem(testKey);
            this._storageAvailable = true;
        } catch (e) {
            this._storageAvailable = false;
        }
        return this._storageAvailable;
    }

    _ensureMigrated() {
        if (this._migrated) return;
        this._migrated = true;
        if (!this._checkStorage()) return;
        try {
            const oldRaw = localStorage.getItem(OLD_SAVE_KEY);
            if (oldRaw && !localStorage.getItem(SAVE_KEY)) {
                localStorage.setItem(SAVE_KEY, oldRaw);
                localStorage.removeItem(OLD_SAVE_KEY);
            }
        } catch (e) {
            // localStorage may not be available
        }
    }

    save(state) {
        if (!this._checkStorage()) return false;
        this._ensureMigrated();
        const compressed = this._compress(state);
        try {
            localStorage.setItem(SAVE_KEY, JSON.stringify(compressed));
            return true;
        } catch (e) {
            console.error('Save failed:', e);
            return false;
        }
    }

    load() {
        if (!this._checkStorage()) return null;
        this._ensureMigrated();
        try {
            const raw = localStorage.getItem(SAVE_KEY);
            if (!raw) return null;
            return this._decompress(JSON.parse(raw));
        } catch (e) {
            console.error('Load failed:', e);
            return null;
        }
    }

    hasSave() {
        if (!this._checkStorage()) return false;
        this._ensureMigrated();
        return !!localStorage.getItem(SAVE_KEY);
    }

    deleteSave() {
        if (!this._checkStorage()) return;
        try {
            localStorage.removeItem(SAVE_KEY);
        } catch (e) {}
    }

    exportSave() {
        if (!this._checkStorage()) return null;
        try {
            const raw = localStorage.getItem(SAVE_KEY);
            if (!raw) return null;
            const blob = new Blob([raw], { type: 'application/json' });
            return URL.createObjectURL(blob);
        } catch (e) {
            return null;
        }
    }

    _compress(state) {
        const compressed = { ...state };

        if (compressed.familyLog && compressed.familyLog.length > FAMILY_LOG_MAX) {
            const keep = compressed.familyLog.slice(-FAMILY_LOG_MAX);
            const summary = this._summarizeLog(compressed.familyLog.slice(0, -FAMILY_LOG_MAX));
            compressed.familyLog = keep;
            compressed.familyLogSummary = summary;
        }

        if (compressed.npcMemories) {
            for (const [npcId, mem] of Object.entries(compressed.npcMemories)) {
                if (mem.events && mem.events.length > 20) {
                    mem.events = mem.events.slice(-20);
                }
            }
            const npcIds = Object.keys(compressed.npcMemories);
            if (npcIds.length > NPC_MEMORY_MAX) {
                const sorted = npcIds.sort((a, b) =>
                    (compressed.npcMemories[b].familiarity || 0) -
                    (compressed.npcMemories[a].familiarity || 0)
                );
                const kept = {};
                for (const id of sorted.slice(0, NPC_MEMORY_MAX)) {
                    kept[id] = compressed.npcMemories[id];
                }
                compressed.npcMemories = kept;
            }
        }

        return compressed;
    }

    _decompress(state) {
        return state;
    }

    _summarizeLog(entries) {
        return {
            count: entries.length,
            generations: [...new Set(entries.map(e => e.generation))],
            preview: entries.slice(-3).map(e => e.title)
        };
    }

    createNewSave(createOptions) {
        return {
            version: 2,
            createdAt: Date.now(),
            seed: Math.floor(Math.random() * 999999),
            generation: 1,
            currentCharacter: null,
            familyLog: [],
            familyLogSummary: null,
            familyRep: {},
            cityAffinity: {},
            rooted: false,
            npcMemories: {},
            inheritedSkillRecipes: [],
            inheritedItems: [],
            inheritedKnowledge: [],
            resonanceShards: 0,
            forgedEquipment: [],
            createOptions: createOptions
        };
    }
}
